src/Tasks/AssemblyDependency/ReferenceTable.cs

```diff
@@ -1225,6 +1253,17 @@ internal static void CalculateParentAssemblyDirectories(List<string> parentRefer
             string rawFileNameCandidate,
             Reference reference)
         {
+            bool isImmutableFrameworkReference = false;
+            if (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_7))
+            {
+                // For a path to be an immutable reference, it must be externally resolved and has a FrameworkReferenceName defined.
+                if (assemblyName == null && !string.IsNullOrEmpty(rawFileNameCandidate) && reference.IsPrimary && reference.ExternallyResolved)
```

Comment:
Why does it have to be Primary? Why does assemblyName have to be null?

Reply:
We're basically saying: If we are given a file path and not assembly name (`assemblyName` is not set, `rawFileNameCandidate` is set), and it is a primary (as in passed to RAR in a parameter, not discovered in dependency walk) and externally resolved (doesn't require dependency walk) dependency, we can consider it to be immutable.

I believe `ExternallyResolved` implies `IsPrimary` so it could be removed but I think this way it captures the concept better. In fact, we could remove this condition altogether if we trust nobody but the SDK task sets `FrameworkReferenceName`. It's more like defense-in-depth so the new logic kicks in only in scenarios where we tested it.

Reply:
> passed to RAR in a parameter, not discovered in dependency walk

This is the part I find most confusing. Why should it matter how a reference got to RAR? Either it's immutable or it isn't, but having RAR specifically reference it shouldn't affect whether it's the sort of reference that changes. I can imagine a design in which the SDK passes all of its references to RAR (and since this change was quite impactful, I'd assume that's true, or we at least pretend it's true), but I could also imagine a design in which the SDK passes a few entrypoint references to RAR, and the rest comes indirectly even though they're all immutable and even in fixed positions relative to each other. Why do we care about which design is used?

Reply:
I think the design with SDK passing all SDK assembly references is something that's unlikely to change. It's an important optimization and we would take several steps back perf-wise if we had to crack open SDK assemblies to find the closure.

Conceptually I agree that we're artificially limiting the scenario to the specifics of how the system works today. I see it more like defense-in-depth to avoid accidentally exposing something that we would later regret supporting.